implemented extensions

dncnmcdougall opened the pull request:
Hi @Furkanzmc , sorry for the 3 month(!) delay. Life. 
(It makes me all the more impressive that open source software happens and is supported at all, thank you.)

This is the implementation of items 1-3 in the spec issue #23.

dncnmcdougall committed:
1. id scheme: Adds id_format option.

The id_format option can be either a string or a function.
If it is a string it is treated as a format string for strftime, as is the
current use. (In this case exposing this option is like exposing the
NOTE_ID_STRFTIME_FORMAT.)
If it is a function it is called with the provide parent_id (possibly
nil) and the list of ids. It should return a new id for the new file.

dncnmcdougall committed:
2. get_notes(): documents the get_notes functions.

dncnmcdougall committed:
3. get_tags(): documents the get_tags()

dncnmcdougall committed:
Bonus: adds "contains" function.

Furkanzmc commented on doc/zettelkasten.txt:
```diff
@@ -147,6 +151,20 @@ setup({opts})                                                *zettelkasten-setup
+              and the list of all the note ids, and should return a new id.
+              For example: >
+              {
+                'id_format': function(parent_id, ids)
```
Passing in the IDs for all notes seem a bit excessive. If the function is interested in this, they can just use the public API and get the IDs themselves.

dncnmcdougall replied:
That is fair. I have removed the parsing in of all the ids.

Furkanzmc commented on doc/zettelkasten.txt:
```diff
@@ -86,11 +86,15 @@ COMMANDS                                                  *zettelkasten-commands
     set with |zettelkasten-setup|, it'll change the working directory.
 
+    Parameters: ~
+        {parent_id} (optional, string): the parent id to use when choosing the
```
It's not clear to me what this `parent_id` is and why it is needed.

dncnmcdougall replied:
So, the idea is to provide the `id_format` function some context on where it is generating an id. For the current use of generating ids from timestamps this is irrelevant.  But if one (like me) wanted to use Ludmann's naming scheme where notes are numbered sequentially: 1,2,3 but can have child notes inserted 1, 1a, 1b, 2, 3, then knowing the `parent_id` is helpful for generating new ids. Namely if one wanted to add to 1a, 1b, sequence one would call `:ZkNew 1`.

I have expanded the documentation a bit to try and explain this in a more generic way.

It is also an optional argument so does not change the public API, but it does "pollute" it a bit. 
What do you think?

Furkanzmc commented on doc/zettelkasten.txt:
```diff
@@ -258,6 +278,72 @@ get_toc({note_id}, {format})                               *zettelkasten-get_toc
+    When selectively setting options for the zettelkasten in the markdown
+    ftplugin: >
+
+    if v:lua.zettelkasten.contains(expand('%'))
```
This is a good example, maybe you should add it to `ftplugin/markdown.lua`.

dncnmcdougall replied:
Thank you. I agree, but I am not sure what, if anything, it would replace: 
Should the entire current markdown.lua be wrapped in a `contains`?

Furkanzmc commented on lua/zettelkasten.lua:
```diff
@@ -58,8 +56,22 @@ local function get_all_tags(lookup_tag)
+
+    local id_format = config.get().id_format
+    local format_type = type(id_format)
+    if  format_type == 'string' then
```
```suggestion
    if format_type == 'string' then
```

dncnmcdougall replied:
This is embarising, I forgot again.

Furkanzmc commented on lua/zettelkasten.lua:
```diff
@@ -58,8 +56,22 @@ local function get_all_tags(lookup_tag)
+    local format_type = type(id_format)
+    if  format_type == 'string' then
+        return fn.strftime(id_format)
+    elseif  format_type == 'function' then
```
```suggestion
    elseif format_type == 'function' then
```

Can you run `stylua` on your changes?

Furkanzmc commented on plugin/zettelkasten.lua:
```diff
@@ -29,18 +29,19 @@ vim.api.nvim_create_autocmd({ "BufReadCmd" }, {
     zkbrowse = function()
         vim.cmd("edit zk://browser")
     end,
+    contains = require("zettelkasten").contains,
```
The reason these are added here is for the commands to access above. I don't think we need this one. And in fact, I should remove the other ones now that there's new API for creating commands.

dncnmcdougall replied:
I have merged your change into this branch too.

Furkanzmc reviewed (changes requested):
Thanks for the changes!

dncnmcdougall committed:
Suggested spelling correction.

Co-authored-by: Furkan Üzümcü <furkanuzumcu@gmail.com>

dncnmcdougall committed:
Parial merge of Furkanzmc into 23_extensions, and spelling correction.

dncnmcdougall committed:
Merge branch dncnmcdougall into 23_extensions

dncnmcdougall committed:
Merge of Furkanzmc into branch: spelling and better commands.

dncnmcdougall committed:
Merge branch dncnmcdougall into 23_extensions

dncnmcdougall committed:
Ran stylua

dncnmcdougall committed:
Removes ids argument to id_format. Updates doc.

dncnmcdougall commented:
Ok, thanks for the feedback.
I have checked teh spelling, run stylua and removed the ids argument.
I have also merged the changes you made in master and updated the example of the `contains` function.
Finally I have improved teh `parent_id` documentation.

dncnmcdougall commented:
Also sorry for the poluted commit history. It got tangled when I merged in master, but didn't realise and then redid it...

dncnmcdougall commented:
Hi @Furkanzmc , is there anything more that I need to do for this pull request?

Furkanzmc reviewed (approved):
Thank you. It looks good! Sorry for the late review, work's been crazy busy and I don't get a lot of time for hobby projects lately :(

dncnmcdougall commented:
Hi, thank you! Sorry for my impatience. I know what you mean about work being crazy!
